File: ai_service/symptom_knowledge_base.py
"""Data-driven symptom knowledge base for clinical triage CDS."""

from __future__ import annotations

import json
import re
from functools import lru_cache
from pathlib import Path
from typing import Any

_KB_PATH = Path(__file__).resolve().parent.parent / "data" / "nlp" / "symptom_knowledge_base.json"
_RF_PATH = Path(__file__).resolve().parent.parent / "data" / "nlp" / "red_flags_library.json"


@lru_cache(maxsize=1)
def load_knowledge_base() -> dict[str, Any]:
    if not _KB_PATH.is_file():
        return {"symptoms": [], "scoring": {}}
    with _KB_PATH.open(encoding="utf-8") as handle:
        data = json.load(handle)
    return data if isinstance(data, dict) else {"symptoms": [], "scoring": {}}


@lru_cache(maxsize=1)
def load_red_flags_library() -> dict[str, Any]:
    if not _RF_PATH.is_file():
        return {"red_flags": [], "policy": {}}
    with _RF_PATH.open(encoding="utf-8") as handle:
        data = json.load(handle)
    return data if isinstance(data, dict) else {"red_flags": [], "policy": {}}


def _term_list(symptom: dict[str, Any]) -> list[str]:
    terms: list[str] = []
    for key in ("keywords", "synonyms", "hiligaynon_terms", "filipino_terms"):
        values = symptom.get(key) or []
        if isinstance(values, list):
            terms.extend(str(v).strip().lower() for v in values if str(v).strip())
    name = str(symptom.get("symptom_name") or "").strip().lower()
    if name:
        terms.append(name)
    # Longest first for phrase preference
    return sorted(set(terms), key=len, reverse=True)


@lru_cache(maxsize=1)
def _symptom_index() -> tuple[dict[str, Any], ...]:
    kb = load_knowledge_base()
    indexed: list[dict[str, Any]] = []
    for raw in kb.get("symptoms") or []:
        if not isinstance(raw, dict):
            continue
        terms = _term_list(raw)
        if not terms:
            continue
        indexed.append({**raw, "_match_terms": terms})
    return tuple(indexed)


def _flexible_phrase_hit(hay: str, term: str) -> bool:
    """Match multi-word phrases allowing short function words between tokens."""
    if not term:
        return False
    if term in hay:
        return True
    parts = [p for p in re.split(r"\s+", term) if p]
    if len(parts) < 2:
        return bool(re.search(rf"(?<!\w){re.escape(term)}(?!\w)", hay))
    # Allow up to 2 intervening tokens (e.g. "masakit akon dughan")
    pattern = r"(?<!\w)" + r"(?:\W+\w+){0,2}\W+".join(re.escape(p) for p in parts) + r"(?!\w)"
    return bool(re.search(pattern, hay))


def _term_matches_with_context(hay: str, term: str, symptom: dict[str, Any]) -> bool:
    if " " in term or "-" in term:
        hit = _flexible_phrase_hit(hay, term)
    else:
        hit = bool(re.search(rf"(?<!\w){re.escape(term)}(?!\w)", hay))
    if not hit:
        return False

    name = str(symptom.get("symptom_name") or "").strip().lower()
    if name and name in hay:
        return True

    if name == "acute abdomen":
        return bool(re.search(r"\b(acute|rigid|peritonitis|sudden severe)\b", hay))

    explicit: list[str] = []
    for key in ("keywords", "synonyms", "hiligaynon_terms", "filipino_terms"):
        for t in symptom.get(key) or []:
            t = str(t).strip().lower()
            if t:
                explicit.append(t)
    explicit = list(dict.fromkeys(explicit))

    generic = {"fever", "pain", "cough", "ache", "bleeding", "weakness", "fatigue", "rash", "swelling"}
    if term in explicit and term not in generic and len(term) >= 6:
        return True
    if " " in term and term in explicit:
        return True

    high_acuity = {"appendicitis pain", "pancreatitis pain", "testicular torsion", "sepsis symptoms", "meningitis symptoms"}
    if name in high_acuity and not re.search(r"\b(severe|acute|sudden|worst|rigid|unbearable|grabe)\b", hay):
        return False

    if re.match(r"^(child|infant|pediatric)\s+", name):
        specific = name.split(" ", 1)[1] if " " in name else ""
        if specific and specific not in hay:
            return False

    if name.endswith(" pain"):
        location = name.replace(" pain", "")
        body_map = {
            "abdominal": r"\b(abdomen|abdominal|stomach|belly|tiyan)\b",
            "back": r"\b(back|likod)\b",
            "chest": r"\b(chest|dughan|dibdib)\b",
            "head": r"\b(head|ulo)\b",
            "neck": r"\b(neck|liog|leeg)\b",
        }
        if location in body_map:
            return bool(re.search(body_map[location], hay))
        if location != "chronic" and location not in hay:
            return False

    name_words = name.split()
    term_words = term.split()
    qual_generic = {"fever", "pain", "cough", "bleeding", "ache", "symptoms", "symptom", "severe", "acute", "chronic", "mild", "high", "low", "with"}

    if len(name_words) <= len(term_words):
        return True

    qualifiers = [w for w in set(name_words) - set(term_words) if len(w) >= 4 and w not in qual_generic]
    if not qualifiers:
        return True

    for q in qualifiers:
        if q in hay:
            return True

    if set(qualifiers) & {"infant", "neonatal", "newborn"}:
        if re.search(r"\b(infant|baby|newborn|sanggol)\b", hay):
            return True
    if set(qualifiers) & {"child", "pediatric"}:
        if re.search(r"\b(child|anak|bata)\b", hay):
            return True

    return False


def match_symptoms(text: str, english_text: str = "", extra_terms: list[str] | None = None) -> list[dict[str, Any]]:
    """Match standardized symptoms from free text using KB synonyms/local terms."""
    haystacks = [
        (text or "").lower(),
        (english_text or "").lower(),
        " ".join(extra_terms or []).lower(),
    ]
    hay = " | ".join(h for h in haystacks if h)
    if not hay.strip("| ").strip():
        return []

    matched: list[dict[str, Any]] = []
    seen: set[str] = set()
    for symptom in _symptom_index():
        sid = str(symptom.get("id") or symptom.get("symptom_name") or "")
        if sid in seen:
            continue
        for term in symptom.get("_match_terms") or []:
            if not term or (len(term) < 5 and " " not in term):
                allow_short = {"ubo", "sipon", "lagnat", "hilo", "tae", "dugo", "hapdi", "kapoy", "luya", "ulon", "mata", "dughan"}
                if term not in allow_short:
                    continue
            if not _term_matches_with_context(hay, term, symptom):
                continue
            matched.append(
                    {
                        "id": symptom.get("id"),
                        "symptom_name": symptom.get("symptom_name"),
                        "medical_category": symptom.get("medical_category"),
                        "severity_weight": int(symptom.get("severity_weight") or 0),
                        "emergency_weight": int(symptom.get("emergency_weight") or 0),
                        "urgent_weight": int(symptom.get("urgent_weight") or 0),
                        "danger_sign": bool(symptom.get("danger_sign")),
                        "recommended_action": symptom.get("recommended_action") or "",
                        "matched_term": term,
                        "common_causes": symptom.get("common_causes") or [],
                        "danger_signs": symptom.get("danger_signs") or [],
                    }
                )
                seen.add(sid)
                break
    # Highest severity first
    matched.sort(key=lambda s: int(s.get("severity_weight") or 0), reverse=True)
    return matched[:8]


def scan_red_flags_library(original: str, english: str = "") -> list[dict[str, Any]]:
    """Scan JSON red-flag library with optional mild-exclusion override."""
    lib = load_red_flags_library()
    policy = lib.get("policy") or {}
    allow_mild = bool(policy.get("allow_mild_override", True))
    hay = f"{(original or '').lower()} {(english or '').lower()}".strip()
    if not hay:
        return []

    matched: list[dict[str, Any]] = []
    seen: set[str] = set()
    for flag in lib.get("red_flags") or []:
        fid = str(flag.get("id") or flag.get("name") or "")
        if not fid or fid in seen:
            continue
        patterns = flag.get("patterns") or {}
        all_patterns: list[tuple[str, str]] = []
        for lang in ("english", "hiligaynon", "filipino"):
            for pat in patterns.get(lang) or []:
                p = str(pat).strip().lower()
                if p:
                    all_patterns.append((lang, p))
        all_patterns.sort(key=lambda x: len(x[1]), reverse=True)

        hit_lang = ""
        hit_pat = ""
        for lang, pat in all_patterns:
            if pat in hay:
                hit_lang, hit_pat = lang, pat
                break
        if not hit_pat:
            continue

        if allow_mild:
            mild_hit = False
            for excl in flag.get("mild_exclusions") or []:
                if str(excl).strip().lower() in hay:
                    mild_hit = True
                    break
            # Explicit mild qualifiers near non-critical wording
            if mild_hit or re.search(r"\bmild\b.{0,40}\b(only|muscle|sore)\b", hay):
                # Still keep truly dangerous phrases
                if not any(x in hay for x in ("cannot breathe", "indi makaginhawa", "unconscious", "vomiting blood", "suicidal")):
                    continue

        matched.append(
            {
                "flag_id": fid,
                "flag_name": flag.get("name") or fid,
                "category": flag.get("category") or "",
                "auto_triage": (flag.get("auto_triage") or "EMERGENCY").upper(),
                "severity_points": int(flag.get("severity_points") or 12),
                "clinical_rationale": flag.get("rationale") or "",
                "matched_on": hit_lang,
                "matched_pattern": hit_pat,
                "english_pattern": hit_pat,
                "source": "red_flags_library.json",
            }
        )
        seen.add(fid)
    return matched


def scoring_config() -> dict[str, Any]:
    return dict(load_knowledge_base().get("scoring") or {})


def clear_cache() -> None:
    load_knowledge_base.cache_clear()
    load_red_flags_library.cache_clear()
    _symptom_index.cache_clear()

File: app/core/SymptomKnowledgeBase.php
<?php
/**
 * Data-driven symptom knowledge base + red-flag library for clinical triage CDS.
 */

final class SymptomKnowledgeBase
{
    /** @var array<string, mixed>|null */
    private static ?array $kb = null;

    /** @var array<string, mixed>|null */
    private static ?array $redFlags = null;

    /** @var list<array<string, mixed>>|null */
    private static ?array $symptomIndex = null;

    /** @var array<string, list<string>>|null */
    private static ?array $csvBoosts = null;

    /** @return array<string, mixed> */
    public static function load(): array
    {
        if (self::$kb !== null) {
            return self::$kb;
        }
        $path = BASE_PATH . '/data/nlp/symptom_knowledge_base.json';
        if (!is_readable($path)) {
            self::$kb = ['symptoms' => [], 'scoring' => []];

            return self::$kb;
        }
        $raw = json_decode((string) file_get_contents($path), true);
        self::$kb = is_array($raw) ? $raw : ['symptoms' => [], 'scoring' => []];

        return self::$kb;
    }

    /** @return array<string, mixed> */
    public static function loadRedFlags(): array
    {
        if (self::$redFlags !== null) {
            return self::$redFlags;
        }
        $path = BASE_PATH . '/data/nlp/red_flags_library.json';
        if (!is_readable($path)) {
            self::$redFlags = ['red_flags' => [], 'policy' => []];

            return self::$redFlags;
        }
        $raw = json_decode((string) file_get_contents($path), true);
        self::$redFlags = is_array($raw) ? $raw : ['red_flags' => [], 'policy' => []];

        return self::$redFlags;
    }

    /** @return array<string, mixed> */
    public static function scoringConfig(): array
    {
        $kb = self::load();

        return is_array($kb['scoring'] ?? null) ? $kb['scoring'] : [];
    }

    /** @return list<array<string, mixed>> */
    private static function symptomIndex(): array
    {
        if (self::$symptomIndex !== null) {
            return self::$symptomIndex;
        }
        $index = [];
        foreach ((self::load()['symptoms'] ?? []) as $symptom) {
            if (!is_array($symptom)) {
                continue;
            }
            $terms = [];
            foreach (['keywords', 'synonyms', 'hiligaynon_terms', 'filipino_terms'] as $key) {
                foreach (($symptom[$key] ?? []) as $term) {
                    $t = strtolower(trim((string) $term));
                    if ($t !== '') {
                        $terms[] = $t;
                    }
                }
            }
            $name = strtolower(trim((string) ($symptom['symptom_name'] ?? '')));
            if ($name !== '') {
                $terms[] = $name;
            }
            $terms = array_values(array_unique($terms));
            usort($terms, static fn (string $a, string $b): int => strlen($b) <=> strlen($a));
            if ($terms === []) {
                continue;
            }
            $symptom['_match_terms'] = $terms;
            $index[] = $symptom;
        }

        // Expand match terms from CSV synonym / Hiligaynon / Filipino term banks (no code change needed to add rows)
        $csvBoost = self::loadCsvTermBoosts();
        foreach ($index as &$symptom) {
            $eng = strtolower((string) ($symptom['symptom_name'] ?? ''));
            $sid = strtolower((string) ($symptom['id'] ?? ''));
            $extra = array_merge($csvBoost[$eng] ?? [], $csvBoost[$sid] ?? []);
            if ($extra !== []) {
                $merged = array_values(array_unique(array_merge($symptom['_match_terms'], $extra)));
                usort($merged, static fn (string $a, string $b): int => strlen($b) <=> strlen($a));
                $symptom['_match_terms'] = $merged;
            }
        }
        unset($symptom);

        self::$symptomIndex = $index;

        return self::$symptomIndex;
    }

    /**
     * Map english/concept → local terms from expandable CSVs.
     *
     * @return array<string, list<string>>
     */
    private static function loadCsvTermBoosts(): array
    {
        if (self::$csvBoosts !== null) {
            return self::$csvBoosts;
        }

        $boost = [];
        // Prefer compact language banks first (full corpora remain available for training/admin)
        $files = [
            BASE_PATH . '/data/nlp/hiligaynon_medical_terms.csv',
            BASE_PATH . '/data/nlp/filipino_medical_terms.csv',
            BASE_PATH . '/data/nlp/medical_phrases.csv',
        ];
        foreach ($files as $path) {
            if (!is_readable($path)) {
                continue;
            }
            $handle = fopen($path, 'r');
            if ($handle === false) {
                continue;
            }
            $header = fgetcsv($handle);
            $count = 0;
            while (($row = fgetcsv($handle)) !== false) {
                $data = array_combine(
                    array_map(static fn ($h) => strtolower(trim((string) $h)), $header ?: []),
                    array_map(static fn ($v) => trim((string) $v), $row)
                ) ?: [];
                $local = strtolower((string) (($data['term'] ?? '') ?: ($data['local_term'] ?? '') ?: ($data['phrase'] ?? '')));
                $eng = strtolower((string) (($data['english'] ?? '') ?: ($data['english_term'] ?? '')));
                $concept = strtolower((string) ($data['concept'] ?? ''));
                if ($local === '' || strlen($local) > 60 || ($eng === '' && $concept === '')) {
                    continue;
                }
                foreach (array_filter([$eng, $concept, str_replace('_', ' ', $concept)]) as $key) {
                    if ($key === '') {
                        continue;
                    }
                    $boost[$key][] = $local;
                }
                $count++;
                if ($count >= 4000) {
                    break;
                }
            }
            fclose($handle);
        }
        foreach ($boost as $k => $list) {
            $boost[$k] = array_values(array_unique($list));
        }
        self::$csvBoosts = $boost;

        return self::$csvBoosts;
    }

    /**
     * @param list<string> $extraTerms
     * @return list<array<string, mixed>>
     */
    public static function matchSymptoms(string $text, string $englishText = '', array $extraTerms = []): array
    {
        $hay = strtolower(trim(implode(' | ', array_filter([
            $text,
            $englishText,
            implode(' ', $extraTerms),
        ]))));
        if ($hay === '') {
            return [];
        }

        $matched = [];
        $seen = [];
        foreach (self::symptomIndex() as $symptom) {
            $sid = (string) ($symptom['id'] ?? $symptom['symptom_name'] ?? '');
            if ($sid === '' || isset($seen[$sid])) {
                continue;
            }
            foreach ($symptom['_match_terms'] as $term) {
                if ($term === '' || (strlen($term) < 5 && !str_contains($term, ' '))) {
                    $allowShort = ['ubo', 'sipon', 'lagnat', 'hilo', 'tae', 'dugo', 'hapdi', 'kapoy', 'luya', 'ulon', 'mata', 'dughan'];
                    if (!in_array($term, $allowShort, true)) {
                        continue;
                    }
                }
                if (!self::termMatchesWithContext($hay, $term, $symptom)) {
                    continue;
                }
                $matched[] = [
                    'id'                  => $symptom['id'] ?? '',
                    'symptom_name'        => $symptom['symptom_name'] ?? '',
                    'medical_category'    => $symptom['medical_category'] ?? '',
                    'severity_weight'      => (int) ($symptom['severity_weight'] ?? 0),
                    'emergency_weight'    => (int) ($symptom['emergency_weight'] ?? 0),
                    'urgent_weight'       => (int) ($symptom['urgent_weight'] ?? 0),
                    'danger_sign'         => (bool) ($symptom['danger_sign'] ?? false),
                    'recommended_action'  => (string) ($symptom['recommended_action'] ?? ''),
                    'matched_term'        => $term,
                    'common_causes'       => $symptom['common_causes'] ?? [],
                    'danger_signs'        => $symptom['danger_signs'] ?? [],
                ];
                $seen[$sid] = true;
                break;
            }
        }
        usort($matched, static fn (array $a, array $b): int => ($b['severity_weight'] <=> $a['severity_weight']));

        // Cap overly broad multi-symptom extraction (prevents score inflation)
        return array_slice($matched, 0, 8);
    }

    /**
     * Require qualifier words when a generic term (fever, pain) would otherwise over-match.
     *
     * @param array<string, mixed> $symptom
     */
    private static function termMatchesWithContext(string $hay, string $term, array $symptom): bool
    {
        if (!self::flexiblePhraseHit($hay, $term)) {
            return false;
        }

        $name = strtolower(trim((string) ($symptom['symptom_name'] ?? '')));
        if ($name !== '' && str_contains($hay, $name)) {
            return true;
        }

        if ($name === 'acute abdomen') {
            return (bool) preg_match('/\b(acute|rigid|peritonitis|sudden severe)\b/u', $hay);
        }

        $explicit = [];
        foreach (['keywords', 'synonyms', 'hiligaynon_terms', 'filipino_terms'] as $key) {
            foreach (($symptom[$key] ?? []) as $t) {
                $t = strtolower(trim((string) $t));
                if ($t !== '') {
                    $explicit[] = $t;
                }
            }
        }
        $explicit = array_values(array_unique($explicit));

        $generic = ['fever', 'pain', 'cough', 'ache', 'bleeding', 'weakness', 'fatigue', 'rash', 'swelling'];
        $stopWords = ['with', 'in', 'and', 'the', 'of', 'a', 'for', 'to', 'sore', 'mild', 'severe'];

        $isLocalTerm = false;
        foreach (['hiligaynon_terms', 'filipino_terms'] as $key) {
            foreach (($symptom[$key] ?? []) as $local) {
                if (strtolower(trim((string) $local)) === $term) {
                    $isLocalTerm = true;
                    break 2;
                }
            }
        }

        // Multi-word symptom names must not match on a single shared English token only.
        if (!$isLocalTerm && str_contains($name, ' ') && $term !== $name) {
            if (!self::compoundQualifiersSatisfied($name, $term, $hay, $generic, $stopWords)) {
                return false;
            }
        }

        if (in_array($term, $explicit, true) && !in_array($term, $generic, true) && strlen($term) >= 6) {
            if ($isLocalTerm || !str_contains($name, ' ') || self::compoundQualifiersSatisfied($name, $term, $hay, $generic, $stopWords)) {
                return true;
            }
        }
        if (str_contains($term, ' ') && in_array($term, $explicit, true)) {
            return true;
        }

        $highAcuity = ['appendicitis pain', 'pancreatitis pain', 'testicular torsion', 'sepsis symptoms', 'meningitis symptoms'];
        if (in_array($name, $highAcuity, true)
            && !preg_match('/\b(severe|acute|sudden|worst|rigid|unbearable|grabe)\b/u', $hay)) {
            return false;
        }

        if (preg_match('/^(child|infant|pediatric)\s+/u', $name, $prefix)) {
            $specific = trim(substr($name, strlen($prefix[0])));
            if ($specific !== '' && !str_contains($hay, $specific)) {
                return false;
            }
        }

        if (str_ends_with($name, ' pain')) {
            $location = str_replace(' pain', '', $name);
            $bodyMap = [
                'abdominal' => '/\b(abdomen|abdominal|stomach|belly|tiyan)\b/u',
                'back' => '/\b(back|likod)\b/u',
                'chest' => '/\b(chest|dughan|dibdib)\b/u',
                'head' => '/\b(head|ulo)\b/u',
                'neck' => '/\b(neck|liog|leeg)\b/u',
            ];
            if (isset($bodyMap[$location])) {
                return (bool) preg_match($bodyMap[$location], $hay);
            }
            if ($location !== 'chronic' && !str_contains($hay, $location)) {
                return false;
            }
        }

        $nameWords = preg_split('/\s+/u', $name) ?: [];
        $termWords = preg_split('/\s+/u', $term) ?: [];
        $qualGeneric = ['fever', 'pain', 'cough', 'bleeding', 'ache', 'symptoms', 'symptom', 'severe', 'acute', 'chronic', 'mild', 'high', 'low', 'with'];

        if (count($nameWords) <= count($termWords)) {
            return true;
        }

        $qualifiers = array_values(array_filter(
            array_diff($nameWords, $termWords),
            static fn (string $w): bool => strlen($w) >= 4 && !in_array($w, $qualGeneric, true)
        ));

        if ($qualifiers === []) {
            return true;
        }

        foreach ($qualifiers as $q) {
            if (str_contains($hay, $q)) {
                return true;
            }
        }

        if (array_intersect($qualifiers, ['infant', 'neonatal', 'newborn'])) {
            if (preg_match('/\b(infant|baby|newborn|sanggol)\b/u', $hay)) {
                return true;
            }
        }
        if (array_intersect($qualifiers, ['child', 'pediatric'])) {
            if (preg_match('/\b(child|anak|bata)\b/u', $hay)) {
                return true;
            }
        }

        return false;
    }

    /** @return list<array<string, mixed>> */
    public static function scanRedFlagsLibrary(string $original, string $english = ''): array
    {
        $lib = self::loadRedFlags();
        $allowMild = (bool) (($lib['policy']['allow_mild_override'] ?? true));
        $hay = strtolower(trim($original . ' ' . $english));
        if ($hay === '') {
            return [];
        }

        $matched = [];
        $seen = [];
        foreach (($lib['red_flags'] ?? []) as $flag) {
            if (!is_array($flag)) {
                continue;
            }
            $fid = (string) ($flag['id'] ?? $flag['name'] ?? '');
            if ($fid === '' || isset($seen[$fid])) {
                continue;
            }
            $patterns = [];
            foreach (['english', 'hiligaynon', 'filipino'] as $lang) {
                foreach (($flag['patterns'][$lang] ?? []) as $pat) {
                    $p = strtolower(trim((string) $pat));
                    if ($p !== '') {
                        $patterns[] = [$lang, $p];
                    }
                }
            }
            usort($patterns, static fn (array $a, array $b): int => strlen($b[1]) <=> strlen($a[1]));

            $hitLang = '';
            $hitPat = '';
            foreach ($patterns as [$lang, $pat]) {
                if (self::redFlagPatternMatches($hay, $pat)) {
                    $hitLang = $lang;
                    $hitPat = $pat;
                    break;
                }
            }
            if ($hitPat === '') {
                continue;
            }

            if ($allowMild) {
                $mildHit = false;
                foreach (($flag['mild_exclusions'] ?? []) as $excl) {
                    if (str_contains($hay, strtolower(trim((string) $excl)))) {
                        $mildHit = true;
                        break;
                    }
                }
                if ($mildHit) {
                    $hard = str_contains($hay, 'cannot breathe')
                        || str_contains($hay, 'indi makaginhawa')
                        || str_contains($hay, 'unconscious')
                        || str_contains($hay, 'vomiting blood')
                        || str_contains($hay, 'suicidal');
                    if (!$hard) {
                        continue;
                    }
                }
            }

            $matched[] = [
                'flag_id'            => $fid,
                'flag_name'          => (string) ($flag['name'] ?? $fid),
                'category'           => (string) ($flag['category'] ?? ''),
                'auto_triage'        => strtoupper((string) ($flag['auto_triage'] ?? 'EMERGENCY')),
                'severity_points'     => (int) ($flag['severity_points'] ?? 12),
                'clinical_rationale' => (string) ($flag['rationale'] ?? ''),
                'matched_on'         => $hitLang,
                'matched_pattern'    => $hitPat,
                'english_pattern'    => $hitPat,
                'source'             => 'red_flags_library.json',
            ];
            $seen[$fid] = true;
        }

        return $matched;
    }

    private static function redFlagPatternMatches(string $hay, string $pattern): bool
    {
        if ($pattern === '') {
            return false;
        }
        // Short patterns (e.g. "PE") must be whole-word matches to avoid false positives in "persistent".
        if (strlen($pattern) <= 3) {
            return (bool) preg_match('/(?<!\w)' . preg_quote($pattern, '/') . '(?!\w)/iu', $hay);
        }

        return str_contains($hay, $pattern);
    }

    /**
     * Require distinguishing qualifiers for compound symptom names.
     *
     * @param list<string> $generic
     * @param list<string> $stopWords
     */
    private static function compoundQualifiersSatisfied(
        string $name,
        string $term,
        string $hay,
        array $generic,
        array $stopWords
    ): bool {
        if (str_contains($hay, $name)) {
            return true;
        }

        // "X with Y" — Y must appear in the complaint (e.g. sore throat with fever → fever required).
        if (preg_match('/\bwith\s+(\w+(?:\s+\w+)?)\b/u', $name, $withMatch)) {
            $required = strtolower(trim($withMatch[1]));
            if ($required !== '' && !self::qualifierPresent($hay, $required)) {
                return false;
            }
        }

        // "X in pregnancy" — pregnancy context required.
        if (str_contains($name, 'pregnancy') && !preg_match('/\b(pregnan|buntis|gravid)\b/u', $hay)) {
            return false;
        }

        $nameWords = preg_split('/\s+/u', $name) ?: [];
        $termWords = preg_split('/\s+/u', $term) ?: [];
        $qualGeneric = ['fever', 'pain', 'cough', 'bleeding', 'ache', 'symptoms', 'symptom', 'severe', 'acute', 'chronic', 'mild', 'high', 'low', 'with', 'in', 'and', 'the', 'of'];

        $required = array_values(array_filter(
            array_diff($nameWords, $termWords),
            static fn (string $w): bool => !in_array($w, array_merge($qualGeneric, $stopWords), true)
                && strlen($w) >= 4
        ));

        if ($required === []) {
            // Still require clinically meaningful qualifiers (swelling, fever, etc.).
            foreach ($nameWords as $w) {
                if (in_array($w, $generic, true) && !in_array($w, $termWords, true)) {
                    $required[] = $w;
                }
            }
        }

        foreach ($required as $q) {
            if (!self::qualifierPresent($hay, $q)) {
                return false;
            }
        }

        return true;
    }

    private static function qualifierPresent(string $hay, string $qualifier): bool
    {
        if (str_contains($hay, $qualifier)) {
            return true;
        }

        $synonyms = [
            'swelling' => '/\b(swollen|hubag|gahabok|edema|pamamaga)\b/u',
            'fever'    => '/\b(lagnat|hilanat|pyrexia|hyperthermia|nilalagnat|ginakalagnat)\b/u',
            'severe'   => '/\b(grabe|worst|unbearable|8\/10|9\/10|10\/10)\b/u',
            'pregnancy'=> '/\b(buntis|gravid)\b/u',
        ];

        return isset($synonyms[$qualifier]) && (bool) preg_match($synonyms[$qualifier], $hay);
    }

    private static function flexiblePhraseHit(string $hay, string $term): bool
    {
        if ($term === '') {
            return false;
        }
        if (str_contains($hay, $term)) {
            return true;
        }
        $parts = preg_split('/\s+/u', $term) ?: [];
        $parts = array_values(array_filter($parts, static fn (string $p): bool => $p !== ''));
        if (count($parts) < 2) {
            return (bool) preg_match('/(?<!\w)' . preg_quote($term, '/') . '(?!\w)/u', $hay);
        }
        $escaped = array_map(static fn (string $p): string => preg_quote($p, '/'), $parts);
        $pattern = '/(?<!\w)' . implode('(?:\W+\w+){0,2}\W+', $escaped) . '(?!\w)/u';

        return (bool) preg_match($pattern, $hay);
    }

    public static function clearCache(): void
    {
        self::$kb = null;
        self::$redFlags = null;
        self::$symptomIndex = null;
        self::$csvBoosts = null;
    }
}
